### ISSUE TITLE
can't upload db by redis-cli --pipe

### ISSUE THREAD

**Issue by rnz** [dragonflydb/dragonfly#4213]
**Describe the bug**

```
time redis-cli -2 --pipe < db0-31073609.resp

real	0m4.453s
user	0m0.355s
sys	0m2.143s
```

```dragonfly_connection.cc:1030] Pipeline buffer over limit: pipeline_bytes 78768428 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit```

but in flagfile:
```
--pipeline_buffer_limit=7008822000
--pipeline_queue_limit=100000
```

**Environment (please complete the following information):**
 - OS: Debian 12
 - Kernel: `Linux n02-test-podman 6.8.12-4-pve #1 SMP PREEMPT_DYNAMIC PMX 6.8.12-4 (2024-11-06T15:04Z) x86_64 GNU/Linux`
 - Containerized: podman
 - "CreateCommand": [
               "podman",
               "run",
               "-dt",
               "--pod",
               "new:dfly",
               "--name",
               "df01",
               "-p",
               "6379:6379",
               "-v",
               "/var/lib/dragonfly:/data",
               "-v",
               "/opt/dfly/etc:/etc/dfly",
               "docker.dragonflydb.io/dragonflydb/dragonfly:latest",
               "--flagfile=/etc/dfly/flagfile.conf"
          ],
 - Dragonfly Version: 1.25.3

**Comment by romange** [comment]
it's just a warning. have you actually saw it does not upload?

**Comment by rnz** [comment]
> it's just a warning. have you actually saw it does not upload?
```
# Keyspace
db0:keys=2293351,expires=0,avg_ttl=-1
```
resp file have 31073609 keys

**Comment by romange** [comment]
Maybe it has duplicates? Can you share the file with me?
this is what I did:
```
printf "set KEY_%d value\n" $(seq 0 31073609) > 1.resp
redis-cli -2 --pipe < 1.resp
All data transferred. Waiting for the last reply...
Last reply received from server.
errors: 0, replies: 31073610
```

```
> dbsize
(integer) 31073610
```

**Comment by romange** [comment]
have not heard from you. so far i have not seen any evidence that there is a bug. closing, please reopen if you have more info

**Comment by romange** [comment]
> @romange
> 
> > haven't heard from you. I haven't seen any evidence that there is a bug yet. closing, please reopen if you have more info
> 
> Well, I'm not a fast Gonzalez, rather a slow Rodriguez )
> 
Had to google these references to understand :)

> printf generates a non resp file. But yes, this file is fully loaded However, the file dumped from keydb is not loaded in dragonfly. But it can be loaded in keydb, redis and garnet.
> 
> Unfortunately, I can't provide the file itself.

Can you please run dragonfly with `--vmodule=dragonfly_connection=1` and provide the info log after the the resp file has finished uploading?

> 
> I also noticed that keys are scanned slowly (generated by printf):
> 
> ```
> # time redis-cli --scan | (pv -p --timer --rate --bytes )| wc -l
>  374MiB 0:06:05 [1.02MiB/s] [                                                    <=>                         ]
> 31073610
> 
> real	6m5.862s
> user	0m14.719s
> sys	3m28.920s
> ```

**Comment by rnz** [comment]
@romange 
> Can you please run dragonfly with `--vmodule=dragonfly_connection=1` and provide the info log after the the resp file has finished uploading?

`--vmodule=dragonfly_connection=1`: 
```
I20241205 15:46:12.598796    10 dragonfly_connection.cc:662] [5] HandleRequests
I20241205 15:46:12.598882    10 dragonfly_connection.cc:1668] [5] LaunchDispatchFiberIfNeeded 
W20241205 15:46:12.702705    10 dragonfly_connection.cc:1030] Pipeline buffer over limit: pipeline_bytes 78768428 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
I20241205 15:46:16.900218    10 dragonfly_connection.cc:984] Error parser status 2
I20241205 15:46:16.900285    10 dragonfly_connection.cc:754] Closed connection for peer id=5 addr=10.88.0.1:59094 laddr=10.88.0.39:6379 fd=676 name= tid=4 irqmatch=1 age=4 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=
```


`--vmodule=dragonfly_connection=2`: 
```
I20241205 15:49:09.352258     6 dragonfly_connection.cc:662] [1] HandleRequests
I20241205 15:49:09.352349     6 dragonfly_connection.cc:1668] [1] LaunchDispatchFiberIfNeeded 
I20241205 15:49:09.352470     6 dragonfly_connection.cc:121] Grown io_buf to 512
I20241205 15:49:09.352504     6 dragonfly_connection.cc:121] Grown io_buf to 1024
I20241205 15:49:09.353669     6 dragonfly_connection.cc:121] Grown io_buf to 2048
I20241205 15:49:09.353709     6 dragonfly_connection.cc:121] Grown io_buf to 4096
W20241205 15:49:09.454347     6 dragonfly_connection.cc:1030] Pipeline buffer over limit: pipeline_bytes 78768428 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
I20241205 15:49:09.801020     6 dragonfly_connection.cc:121] Grown io_buf to 8192
I20241205 15:49:11.235170     6 dragonfly_connection.cc:121] Grown io_buf to 32768
I20241205 15:49:11.620589     6 dragonfly_connection.cc:121] Grown io_buf to 65536
I20241205 15:49:13.765785     6 dragonfly_connection.cc:969] Before dispatch_fb.join()
I20241205 15:49:13.765923     6 dragonfly_connection.cc:971] After dispatch_fb.join()
I20241205 15:49:13.768098     6 dragonfly_connection.cc:984] Error parser status 2
I20241205 15:49:13.768144     6 dragonfly_connection.cc:754] Closed connection for peer id=1 addr=10.88.0.1:34728 laddr=10.88.0.40:6379 fd=676 name= tid=0 irqmatch=1 age=4 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=
```

**Comment by romange** [comment]
yeah, please run dragonfly with `--max_multi_bulk_len=1000000` and tell me if it helps.
I will make the warning more user friendly

**Comment by rnz** [comment]
@romange `--max_multi_bulk_len=1000000` doesn't help too:
```
I20241205 18:26:37.683950     6 dragonfly_connection.cc:662] [1] HandleRequests
I20241205 18:26:37.684032     6 dragonfly_connection.cc:1668] [1] LaunchDispatchFiberIfNeeded 
I20241205 18:26:37.684060     6 dragonfly_connection.cc:121] Grown io_buf to 512
I20241205 18:26:37.684079     6 dragonfly_connection.cc:121] Grown io_buf to 1024
I20241205 18:26:37.685305     6 dragonfly_connection.cc:121] Grown io_buf to 2048
I20241205 18:26:37.685393     6 dragonfly_connection.cc:121] Grown io_buf to 4096
W20241205 18:26:37.790148     6 dragonfly_connection.cc:1030] Pipeline buffer over limit: pipeline_bytes 78768428 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
I20241205 18:26:38.147810     6 dragonfly_connection.cc:121] Grown io_buf to 8192
I20241205 18:26:39.554211     6 dragonfly_connection.cc:121] Grown io_buf to 32768
I20241205 18:26:39.922112     6 dragonfly_connection.cc:121] Grown io_buf to 65536
I20241205 18:26:42.053949     6 dragonfly_connection.cc:969] Before dispatch_fb.join()
I20241205 18:26:42.054078     6 dragonfly_connection.cc:971] After dispatch_fb.join()
I20241205 18:26:42.056334     6 dragonfly_connection.cc:984] Error parser status 2
I20241205 18:26:42.056417     6 dragonfly_connection.cc:754] Closed connection for peer id=1 addr=10.88.0.1:45752 laddr=10.88.0.43:6379 fd=676 name= tid=0 irqmatch=1 age=5 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=
```

**Comment by rnz** [comment]
@romange: 
```
# podman logs -f df01
I20241212 16:56:22.932678     1 init.cc:78] dragonfly running in opt mode.
* Logs will be written to the first available of the following paths:
/tmp/dragonfly.*
./dragonfly.*
* For the available flags type dragonfly [--help | --helpfull]
* Documentation can be found at: https://www.dragonflydb.io/docs
I20241212 16:56:22.932905     1 dfly_main.cc:693] Starting dragonfly df-v1.25.5-ad3e037555b61889e81b7e072dd0896f627bc90e
I20241212 16:56:22.933216     1 dfly_main.cc:737] maxmemory has not been specified. Deciding myself....
I20241212 16:56:22.933225     1 dfly_main.cc:746] Found 1.44TiB available memory. Setting maxmemory to 1.15TiB
W20241212 16:56:22.933301     1 dfly_main.cc:363] iouring API is not supported. switching to epoll.
I20241212 16:56:23.056463     1 proactor_pool.cc:147] Running 336 io threads
I20241212 16:56:23.110875     1 server_family.cc:835] Host OS: Linux 6.8.12-5-pve x86_64 with 336 threads
I20241212 16:56:23.145934     1 snapshot_storage.cc:181] Load snapshot: Searching for snapshot in directory: "/data"
I20241212 16:56:23.153913     1 server_family.cc:1102] Loading /data/dump-2024-12-12T16:56:07-summary.dfs
I20241212 16:56:23.215654     7 listener_interface.cc:101] sock[675] AcceptServer - listening on port 6379
I20241212 16:56:23.747960    10 server_family.cc:1155] Load finished, num keys read: 31073610
I20241212 16:57:08.344563     6 dragonfly_connection.cc:662] [1] HandleRequests
W20241212 16:57:08.344920     6 main_service.cc:1410]  COMMAND DOCS failed with reason: syntax error
I20241212 16:57:44.471609     6 save_stages_controller.cc:337] Saving "dump-2024-12-12T16:57:44-summary.dfs" finished after 0 us
I20241212 16:57:45.428550     6 dragonfly_connection.cc:1203] [1] Got event 8197 1 0 system:0
I20241212 16:57:45.428622     6 dragonfly_connection.cc:969] Before dispatch_fb.join()
I20241212 16:57:45.428630     6 dragonfly_connection.cc:971] After dispatch_fb.join()
I20241212 16:57:45.428653     6 dragonfly_connection.cc:754] Closed connection for peer id=1 addr=10.88.0.1:34760 laddr=10.88.0.8:6379 fd=676 name= tid=0 irqmatch=0 age=37 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=

# here - try import:
I20241212 16:58:11.322113     7 dragonfly_connection.cc:662] [2] HandleRequests
I20241212 16:58:11.323220     7 dragonfly_connection.cc:1668] [2] LaunchDispatchFiberIfNeeded 
I20241212 16:58:11.323285     7 dragonfly_connection.cc:121] Grown io_buf to 512
I20241212 16:58:11.323395     7 dragonfly_connection.cc:121] Grown io_buf to 1024
I20241212 16:58:11.324515     7 dragonfly_connection.cc:121] Grown io_buf to 2048
I20241212 16:58:11.324539     7 dragonfly_connection.cc:121] Grown io_buf to 4096
W20241212 16:58:11.426340     7 dragonfly_connection.cc:1030] Pipeline buffer over limit: pipeline_bytes 78768428 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
I20241212 16:58:11.769240     7 dragonfly_connection.cc:121] Grown io_buf to 8192
I20241212 16:58:13.186705     7 dragonfly_connection.cc:121] Grown io_buf to 32768
I20241212 16:58:13.570048     7 dragonfly_connection.cc:121] Grown io_buf to 65536
I20241212 16:58:15.690116     7 dragonfly_connection.cc:969] Before dispatch_fb.join()
I20241212 16:58:15.696918     7 dragonfly_connection.cc:971] After dispatch_fb.join()
I20241212 16:58:15.700935     7 dragonfly_connection.cc:984] Error parser status 2
I20241212 16:58:15.700974     7 dragonfly_connection.cc:754] Closed connection for peer id=2 addr=10.88.0.1:57182 laddr=10.88.0.8:6379 fd=676 name= tid=1 irqmatch=0 age=4 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=
```

Import procedure:
```
# time redis-cli -p 7379 -2 --pipe < db0-again-31073609.resp

real	0m4.385s
user	0m0.403s
sys	0m2.415s
```
and very slow scan after:
```
root@n02-test-podman:~# time redis-cli -p 7379 --scan | pv | wc -l
76.7MiB 0:00:29 [2.60MiB/s] [                   <=>             ]
2300838

real	0m29.509s
user	0m1.284s
sys	0m15.539s
```

**Comment by romange** [comment]
This is going to be fixed in 1.26. I can also  trigger a weekly docker build for you to test tomorrow

**Comment by romange** [comment]
@rnz  please try docker.dragonflydb.io/dragonflydb/dragonfly-weekly:ubuntu  with the same flags and tell me how it went

**Comment by rnz** [comment]
@romange 
```
# podman pull docker.dragonflydb.io/dragonflydb/dragonfly-weekly:ubuntu
Trying to pull docker.dragonflydb.io/dragonflydb/dragonfly-weekly:ubuntu...
Error: initializing source docker://docker.dragonflydb.io/dragonflydb/dragonfly-weekly:ubuntu: reading manifest ubuntu in docker.dragonflydb.io/dragonflydb/dragonfly-weekly: name unknown: repository name not known to registry
```
But:
```
# podman pull docker.dragonflydb.io/dragonflydb/dragonfly
Trying to pull docker.dragonflydb.io/dragonflydb/dragonfly:latest...
Getting image source signatures
Copying blob 6b210fb3ff3a skipped: already exists  
Copying blob e717618de34d skipped: already exists  
Copying blob 4f4fb700ef54 skipped: already exists  
Copying blob 093adbc1e584 skipped: already exists  
Copying blob 56f34fb320d5 skipped: already exists  
Copying blob ecf0691fc1f9 skipped: already exists  
Copying blob 63face880dc5 skipped: already exists  
Copying blob 6414378b6477 skipped: already exists  
Copying config c316a36ccb done  
Writing manifest to image destination
Storing signatures
c316a36ccb3a2fa0de65d48d622c624c560b9aa004cbee51baac91b246df1981
```

**Comment by romange** [comment]
sorry, i meant: `docker pull ghcr.io/dragonflydb/dragonfly-weekly:ubuntu`

**Comment by romange** [comment]
@rnz  hi, did it solve the issue?

**Comment by rnz** [comment]
@romange  
I just checked - yes it solve, now import is done:
```
# time redis-cli -p 7379 -2 --pipe < db0-again-31073609.resp
All data transferred. Waiting for the last reply...
Last reply received from server.
errors: 0, replies: 31169367

real	1m0.670s
user	0m5.301s
sys	0m36.883s

# redis-cli -p 7379 info keyspace 
# Keyspace
db0:keys=31073609,expires=0,avg_ttl=-1
````
log when import:
```
2b9a81f2cee3 I20241216 14:31:47.606841     7 dragonfly_connection.cc:662] [2] HandleRequests
2b9a81f2cee3 I20241216 14:31:49.437201     7 save_stages_controller.cc:337] Saving "dump-2024-12-16T14:31:49-summary.dfs" finished after 0 us
2b9a81f2cee3 I20241216 14:31:53.165820     7 dragonfly_connection.cc:1213] [2] Got event 8197 1 0 system:0
2b9a81f2cee3 I20241216 14:31:53.165880     7 dragonfly_connection.cc:979] Before dispatch_fb.join()
2b9a81f2cee3 I20241216 14:31:53.165889     7 dragonfly_connection.cc:981] After dispatch_fb.join()
2b9a81f2cee3 I20241216 14:31:53.165899     7 dragonfly_connection.cc:761] Closed connection for peer id=2 addr=10.88.0.1:57258 laddr=10.88.0.13:6379 fd=676 name= tid=1 irqmatch=0 age=6 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=
2b9a81f2cee3 I20241216 14:31:57.835084     8 dragonfly_connection.cc:662] [3] HandleRequests
2b9a81f2cee3 I20241216 14:31:57.835306     8 dragonfly_connection.cc:1683] [3] LaunchDispatchFiberIfNeeded 
2b9a81f2cee3 I20241216 14:31:57.835418     8 dragonfly_connection.cc:121] Grown io_buf to 512
2b9a81f2cee3 I20241216 14:31:57.835441     8 dragonfly_connection.cc:121] Grown io_buf to 1024
2b9a81f2cee3 I20241216 14:31:57.835610     8 dragonfly_connection.cc:121] Grown io_buf to 2048
2b9a81f2cee3 I20241216 14:31:57.835633     8 dragonfly_connection.cc:121] Grown io_buf to 4096
2b9a81f2cee3 W20241216 14:31:57.940541     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 78768428 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 I20241216 14:31:59.757566     8 dragonfly_connection.cc:121] Grown io_buf to 32768
2b9a81f2cee3 I20241216 14:32:00.121565     8 dragonfly_connection.cc:121] Grown io_buf to 65536
2b9a81f2cee3 W20241216 14:32:02.375715     8 redis_parser.cc:330] Large bulk len: 357914744
2b9a81f2cee3 W20241216 14:32:05.513969     8 redis_parser.cc:330] Large bulk len: 357914756
2b9a81f2cee3 W20241216 14:32:08.015504     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 875489796 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 W20241216 14:32:08.891621     8 redis_parser.cc:330] Large bulk len: 357914776
2b9a81f2cee3 W20241216 14:32:18.042801     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 889437155 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 W20241216 14:32:28.107385     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 890490695 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 W20241216 14:32:34.920727     8 redis_parser.cc:330] Large bulk len: 357914844
2b9a81f2cee3 W20241216 14:32:38.131341     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 896907547 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 W20241216 14:32:42.798081     8 redis_parser.cc:330] Large bulk len: 357914840
2b9a81f2cee3 W20241216 14:32:46.056071     8 redis_parser.cc:330] Large bulk len: 357914824
2b9a81f2cee3 W20241216 14:32:48.131729     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 883226814 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 W20241216 14:32:58.157135     8 dragonfly_connection.cc:1040] Pipeline buffer over limit: pipeline_bytes 891815063 queue_size 100001, consider increasing pipeline_buffer_limit/pipeline_queue_limit
2b9a81f2cee3 I20241216 14:32:58.501397     8 dragonfly_connection.cc:1213] [3] Got event 8197 1 0 system:0
2b9a81f2cee3 I20241216 14:32:58.501452     8 dragonfly_connection.cc:979] Before dispatch_fb.join()
2b9a81f2cee3 I20241216 14:32:58.501462     8 dragonfly_connection.cc:981] After dispatch_fb.join()
2b9a81f2cee3 I20241216 14:32:58.501469     8 dragonfly_connection.cc:761] Closed connection for peer id=3 addr=10.88.0.1:41436 laddr=10.88.0.13:6379 fd=676 name= tid=2 irqmatch=0 age=61 idle=0 db=0 flags=t phase=preclose lib-name= lib-ver=
```
